FEA Add Rand Index and pair confusion matrix

ufmayer opened the pull request:
<!--
Thanks for contributing a pull request! Please ensure you have taken a look at
the contribution guidelines: https://github.com/scikit-learn/scikit-learn/blob/master/CONTRIBUTING.md#pull-request-checklist
-->

#### Reference Issues/PRs
Adding new functionality


#### What does this implement/fix? Explain your changes.
Two changes:
1. contingency_matrix() creates currently an int overflow for large data such as using adjusted_rand_score on large clusterings. This implementation adds an optional dtype parameter which can be used to store the counts in np.int64 to avoid overflows.
2. Adds computation of the unadjusted Rand Index (https://en.wikipedia.org/wiki/Rand_index).  Computation is based on an efficient pair comparison helper method similar to the existing
adjusted_rand_score method.

#### Any other comments?
Rand Index is essentially computing the non-chance-adjusted accuracy of a clustering, which is a standard measure for clusterings.

ufmayer committed:
Add dtype arg to contingency_matrix to avoid int overflow for larger matrices

ufmayer committed:
Add pair_confusion_matrix and rand_score

jnothman commented:
Thanks for the PR.
The linter is unhappy.

At some point (I can't find the issue) I proposed that we have a function which turns a binary classification metric into a clustering metric via the pairwise transformation. In that framework, this would just be accuracy... Not sure if we want to go about implementing the generic solution. I know that in some spaces, F1 at least is calculated on pairs of clustered elements.

(Another generic class of clustering metrics is the max-sum assignment, which here we only implement as the consensus score for biclustering.)

jnothman commented on sklearn/metrics/cluster/_supervised.py:
```diff
@@ -138,6 +145,209 @@ def contingency_matrix(labels_true, labels_pred, *, eps=None, sparse=False):
+    D10 = sum(contingency.transpose().dot(n_c)) - Sum_Squares
+    #D00 = sum(contingency[c,k] * (n_samples - n_c[c] - n_k[k] + contingency[c,k]) for c,k in zip(rows,cols))
+    D00 = n_samples * n_samples - D01 - D10 - Sum_Squares
+    return [[D00, D01], [D10, D11]]
```
please return an array

jnothman commented on sklearn/metrics/cluster/_supervised.py:
```diff
@@ -138,6 +145,209 @@ def contingency_matrix(labels_true, labels_pred, *, eps=None, sparse=False):
 
 # clustering measures
 
+def pair_confusion_matrix(labels_true, labels_pred):
```
this function needs tests in `sklearn/metrics/cluster/tests/test_supervised.py`

jnothman commented on sklearn/metrics/cluster/_supervised.py:
```diff
@@ -138,6 +145,209 @@ def contingency_matrix(labels_true, labels_pred, *, eps=None, sparse=False):
+    #D11 = sum(n_ij*(n_ij-1) for n_ij in contingency.data)
+    Sum_Squares = sum(n_ij*n_ij for n_ij in contingency.data)
+    D11 = Sum_Squares - n_samples
+    #D01 = sum(contingency[c,k] * (n_k[k] - contingency[c,k]) for c,k in zip(rows,cols))
```
I don't think we need these comments.

jnothman commented on sklearn/metrics/cluster/_supervised.py:
```diff
@@ -138,6 +145,209 @@ def contingency_matrix(labels_true, labels_pred, *, eps=None, sparse=False):
+    Sum_Squares = sum(n_ij*n_ij for n_ij in contingency.data)
+    D11 = Sum_Squares - n_samples
+    #D01 = sum(contingency[c,k] * (n_k[k] - contingency[c,k]) for c,k in zip(rows,cols))
+    D01 = sum(contingency.dot(n_k)) - Sum_Squares
```
Is this a marked improvement than the implementation of the pairwise confusion in adjusted_rand_index using `_comb2`?

ufmayer replied:
Yes, for larger clusterings it runs quite a bit faster. A _comb2 implementation of the un-adjusted rand score might be:
```
    sum_comb = sum(_comb2(n_ij) for n_ij in contingency.data)
    sum_comb_c = sum(_comb2(n_c) for n_c in np.ravel(contingency.sum(axis=1)))
    sum_comb_k = sum(_comb2(n_k) for n_k in np.ravel(contingency.sum(axis=0)))
    D11 = 2 * sum_comb
    D01 = 2 * (sum_comb_k-sum_comb)
    D10 = 2 * (sum_comb_c-sum_comb)
    D00 = n_samples * n_samples - n_samples - D01 - D10 - D11
```
A runtime test case can be generated with:
```
n = 10000
clustering1 = list(range(n))
clustering2 = list(range(n))
# randomly create consecutive pairs for 5% of the entries
np.random.seed(12345)
clustering1_pairs = sorted(np.random.choice(clustering1[1:], size=int(0.05*n), replace=False))
for i in clustering1_pairs:
    clustering1[i-1] = clustering1[i]
clustering2_pairs = sorted(np.random.choice(clustering2[1:], size=int(0.05*n), replace=False))
for i in clustering2_pairs:
    clustering2[i-1] = clustering2[i]
```
On my machine %timeit for the two implementation returns:
```
pair_confusion_matrix(clustering1, clustering2)
10.3 ms ± 530 µs per loop (mean ± std. dev. of 7 runs, 100 loops each)
pair_confusion_matrix_comb2(clustering1, clustering2)
20.2 ms ± 735 µs per loop (mean ± std. dev. of 7 runs, 100 loops each)
```
Run-time differences highly depend on the structure of the clusterings. I have not seen a case where my implementation is slower.

jnothman commented on sklearn/metrics/cluster/_supervised.py:
```diff
@@ -138,6 +145,209 @@ def contingency_matrix(labels_true, labels_pred, *, eps=None, sparse=False):
+    return [[D00, D01], [D10, D11]]
+
+
+def rand_score(labels_true, labels_pred):
```
this function needs tests in `sklearn/metrics/cluster/tests/test_supervised.py`, and inclusion in `sklearn/metrics/cluster/tests/test_common.py`

ufmayer committed:
Added tests for rand_score and pair_confusion_matrix

ufmayer committed:
Changed return type of pair_confusion_matrix from list to array

ufmayer committed:
Removed some superfluous code comments from pair_confusion_matrix

ufmayer committed:
Remove doctest: +ELLIPSIS from comments

ufmayer committed:
Break up long lines

ufmayer committed:
Break up long lines

ufmayer committed:
Remove duplicate imports

ufmayer committed:
Whitespace changes for linter

ufmayer committed:
Another whitespace change for linter

ufmayer commented:
Currently it passes the linter but it fails to fully build because the test code is looking for the new methods to be part of the package, but as they are not yet merged they are not found. That seems like a generic chicken-and-egg problem, but maybe I am misunderstanding something on how this is set up.
```
sklearn/tests/test_supervised.py:6: error: Module 'sklearn.metrics' has no attribute 'pair_confusion_matrix'
sklearn/tests/test_supervised.py:6: error: Module 'sklearn.metrics' has no attribute 'rand_score' sklearn/metrics/cluster/tests/test_common.py:8: error: Module 'sklearn.metrics.cluster' has no attribute 'rand_score'
```

ufmayer commented:
@jnothman I made the requested changes. Haven't seen any update on this for a week. I don't know how fast these things usually progress.l Is there something else that's waited for from me?

cmarmo commented:
Hi @ufmayer, thanks for your work. Some checks didn't pass, do you mind to have a look? Thanks!

cmarmo commented:
Please, forgive me @ufmayer you already noticed the unsuccessful checks, my comment wasn't helpful.
The `mypy` [check](https://dev.azure.com/scikit-learn/scikit-learn/_build/results?buildId=17338&view=logs&j=32e2e1bb-a28f-5b18-6cfc-3f01273f5609&t=8a54543f-0728-5134-6642-bedd98e03dd0) tells you that `rand_score` and `pair_confusion_matrix` should be defined in the [`sklearn/metrics/cluster/__init__.py`](https://github.com/scikit-learn/scikit-learn/blob/master/sklearn/metrics/cluster/__init__.py) to be imported correctly. Doing that will let the build know where the new module is.
Hope this would help to move forward.

ufmayer committed:
Fixed bugs in test_supervised

ufmayer committed:
Merged tests/test_supervised.py into metrics/cluster/tests/test_supervised.py

ufmayer committed:
Added rand_score and pair_confusion_matrix to various init, test, and scorer files

ufmayer committed:
added rand_score and pair_confusion_matrix to __all__ in __init__

ufmayer committed:
changed whitespace to make linter happy

ufmayer committed:
Update comments in pair_confusion_matrix for doctest

ufmayer commented:
@cmarmo Merci beaucoup. That was the information I needed. All checks have now passed. What's the next step?

cmarmo commented:
@ufmayer, je vous en prie... :)

> What's the next step?

'Waiting for reviewer' ... :) , pinging from time to time for the pull request don't get lost. Also it is useful to check if conflicts are generated by other merge, and fix them when they arise.
Last suggestion: you can add [MRG] to the PR title (as specified in the [documentation](https://scikit-learn.org/stable/developers/contributing.html#pull-request-checklist)), to specify that your PR is ready for review. Good luck!

jnothman commented:
Sorry, my availability lately has been very scattered. I commented on this
because I was interested in it, not because I had excess time for review!
:D

I'll try give it another shot.

ufmayer commented:
@jnothman Thanks for saying you might find time to give this another shot. I think it's ready to go. How about it?

ufmayer committed:
Merged in master from from the upstream repository

ufmayer committed:
doc changes after merging in upstream master

ufmayer committed:
more doc changes after merging in upstream master

jnothman reviewed (commented):
> @jnothman Thanks for saying you might find time to give this another shot. I think it's ready to go. How about it?

I reviewed this weeks ago, but my review didn't submit, apparently. Then I tried to submit it the other day, but apparently no luck there. I wonder how many of my reviews are pending!

ufmayer committed:
Added myself to authors

ufmayer committed:
Whitespace changes

ufmayer commented on sklearn/metrics/cluster/_supervised.py:
```diff
@@ -138,6 +145,211 @@ def contingency_matrix(labels_true, labels_pred, *, eps=None, sparse=False):
+
+    Examples
+    --------
+
```
Done. I did the same change for ARI.

ufmayer commented on sklearn/metrics/cluster/_supervised.py:
```diff
@@ -138,6 +145,211 @@ def contingency_matrix(labels_true, labels_pred, *, eps=None, sparse=False):
+
+    References
+    ----------
+
```
Done. I did the same change for ARI.

ufmayer commented on sklearn/metrics/cluster/_supervised.py:
```diff
@@ -138,6 +145,211 @@ def contingency_matrix(labels_true, labels_pred, *, eps=None, sparse=False):
+    diagonal regardless of actual label values:
+
+      >>> from sklearn.metrics.cluster import pair_confusion_matrix
+      >>> pair_confusion_matrix([0, 0, 1, 1], [0, 0, 1, 1])
```
I followed the documentation in adjusted_rand_index. I think it's quite legible. If this ought to be changed then likely adjusted_rand_index should be changed as well.

ufmayer commented on sklearn/metrics/cluster/tests/test_supervised.py:
```diff
@@ -160,6 +164,11 @@ def test_non_consecutive_labels():
 
+    ri_1 = rand_score([0, 0, 0, 1, 1, 1], [0, 1, 0, 1, 2, 2])
+    ri_2 = rand_score([0, 0, 0, 1, 1, 1], [0, 4, 0, 4, 2, 2])
+    assert_almost_equal(ri_1, 0.66, 2)
```
Yes, this test could likely be generalized and simply compare the results of two identical clusterings that only differ in their label values. On the other hand having an explicit value tests additionally that the routine works as expected (it could theoretically be twice wrong but compute the same values if one test equality on two computations). I have follow the adjusted_rand_score test with my code. I'd suggest to leave this as is and put it up for future restructuring once the rand_index is integrated, as this is a larger issue affecting a more general test setup for several metrics.

jnothman replied:
I parenthesised my comment because it wasn't about your current contribution.

You already test the explicit results, but I think this question of how a metric should behave with a particular anomalous feature of input _should_ be tested through common tests and should _not_ be tested against a be specific result. Again, not your problem

ufmayer commented on sklearn/metrics/cluster/_supervised.py:
```diff
@@ -138,6 +145,211 @@ def contingency_matrix(labels_true, labels_pred, *, eps=None, sparse=False):
+    not clustered together, and D11 = number of pairs with both clusterings
+    having the samples clustered together.
+
+    RI is a symmetric measure::
```
I followed the documentation in adjusted_rand_index. I think it's quite legible. If this ought to be changed then likely adjusted_rand_index should be changed as well. We might want to leave that now and do a more whole-scale cleanup later for both rand indexes.

jnothman replied:
I'm okay having this detail here too, but it needs to be in the narrative documentation i.e. https://scikit-learn.org/stable/modules/clustering.html#clustering-performance-evaluation

ufmayer replied:
Added documentation of the functions to doc/modules/clustering.rst at the appropriate places.

ufmayer committed:
Implement rand_score test as hardcoded example

ufmayer commented on sklearn/metrics/cluster/tests/test_supervised.py:
```diff
@@ -348,3 +357,73 @@ def test_fowlkes_mallows_score_properties():
+    assert_array_equal(expected, pair_confusion_matrix(clustering1,
+                                                       clustering2))
+
+    # regular case: different non-trivial clusterings
```
Done.

ufmayer commented on sklearn/metrics/cluster/_supervised.py:
```diff
@@ -138,6 +145,211 @@ def contingency_matrix(labels_true, labels_pred, *, eps=None, sparse=False):
+    pairs that are assigned in the same or different clusters in the
+    predicted and true clusterings.
+
+    The 2x2 pair confusion matrix is:
```
Apparently there is no generated artifact under ci/circleci doc. I did find the ci/circleci doc artifact for adjusted_rand_score so presumably I was looking in the right place. Does rand_score have to be added somewhere to generate the html doc?

ufmayer replied:
![pair_confusion_matrix_html_rendering](https://user-images.githubusercontent.com/6536841/85911707-7fea3e00-b7db-11ea-8d43-83af9da9d5e0.PNG)

Looks good to me.

ufmayer committed:
Removed special case handling from pair_confusion_matrix, not needed

ufmayer committed:
Change pair_confusion_matrix to use _comb2 for consistency with adjusted_rand_score implmentation

ufmayer commented on sklearn/metrics/cluster/_supervised.py:
```diff
@@ -138,6 +145,211 @@ def contingency_matrix(labels_true, labels_pred, *, eps=None, sparse=False):
+    # These are perfect matches hence all pairs are counted either under
+    # D00 or D11.
+    if n_classes == n_clusters == 1:
+        return [[0, 0], [0, n_samples * (n_samples - 1)]]
```
I had followed the adjusted_rand_score setup, but no, we don't need these. I removed them.

ufmayer committed:
Whitespace change to make linter happy

ufmayer committed:
Removed unused variables left over from code change

ufmayer commented:
@jnothman As per your suggestions I changed the implementation to use _comb2 (like adjusted_rand_score) for ease of maintenance, and replaced the test with a toy example. I removed the special case handling, it’s not needed and again makes the code simpler. I also made the requested whitespace changes, and added myself to the set of authors. I left some things untouched because they duplicate the setup of adjusted_rand_score exactly, please see the comments above. One outstanding issue: I could not find any ci/circleci doc artifacts for the contributed methods, do these methods have to get added somewhere so generation gets triggered?

ufmayer committed:
Added section on Rand Index to docs

ufmayer committed:
Added index entry for Rand Index to docs

ufmayer committed:
Added entry for Rand Index to model evaluation docs

ufmayer committed:
Added rand_score to change log

ufmayer committed:
Add docs for pair_confusion_matrix

ufmayer committed:
Pair confustion matrix comment changes only

ufmayer committed:
More comment and doc changes only

ufmayer committed:
Remove cluster.dbscan covariance.oas decomposition.fastica from function autosummary in classes.rst

ufmayer commented:
@jnothman I added the documentation you requested. After that my builds were failing repeatedly because I had changed the doc/modules/classes.rst file and Sphinx triggered warnings about a different part of the file not related to my changes. It turns out FastICA, DBSCAN, and OAS were in both the classes autosummary and (lowercased) in the function autosummary sections (these are classes). Sphinx complained about the function autosummaries and the build failed:
```
/home/circleci/project/doc/modules/classes.rst.rst:115: WARNING: autosummary:
stub file not found '\''sklearn.cluster.dbscan'\''. Check your autosummary_generate setting.
/home/circleci/project/doc/modules/classes.rst.rst:184: WARNING: autosummary:
stub file not found '\''sklearn.covariance.oas'\''. Check your autosummary_generate setting.
/home/circleci/project/doc/modules/classes.rst.rst:325: WARNING: autosummary:
stub file not found '\''sklearn.decomposition.fastica'\''. Check your autosummary_generate setting.'
```
After I fixed that by removing them from the function autosummary section of classes.rst the build succeeded.

As far as I can tell the PR is now complete. Thank you for your careful review and comments.

ufmayer committed:
Fix a typo in the comments

ufmayer committed:
Merged in master from from the upstream repository

ufmayer commented:
@jnothman I merged in the master again. At this point I find myself chasing the master, a final review of this PR and merge would be nice.

jnothman commented on doc/modules/classes.rst:
```diff
@@ -120,7 +120,6 @@ Functions
    cluster.cluster_optics_dbscan
    cluster.cluster_optics_xi
    cluster.compute_optics_graph
-   cluster.dbscan
```
Please revert these changes. They were issues in master that have been patched there.

ufmayer replied:
Done.

jnothman reviewed (commented):
> At this point I find myself chasing the master, a final review of this PR and merge would be nice.

Mine seems to be the first, not the final, review. We require 2 approvals before merge.

Sorry that it's been slow. Our issue tracker has had a lot of traffic, and I've been trying to work on [big challenges](https://scikit-learn-enhancement-proposals.readthedocs.io/en/latest/) rather than small features. I hope to review this more fully soon.

jnothman commented on sklearn/metrics/cluster/_supervised.py:
```diff
@@ -138,6 +146,206 @@ def contingency_matrix(labels_true, labels_pred, *, eps=None, sparse=False):
+    pairs that are assigned in the same or different clusters in the
+    predicted and true clusterings.
+
+    The 2x2 pair confusion matrix::
```
Please replace this with "read more in the user guide" with the syntax used in most other API docstrings.

ufmayer replied:
Done.

ufmayer replied:
Done.

jnothman commented on sklearn/metrics/cluster/_supervised.py:
```diff
@@ -138,6 +146,206 @@ def contingency_matrix(labels_true, labels_pred, *, eps=None, sparse=False):
+
+    Returns
+    -------
+    nested array [[D00, D01], [D10, D11]]
```
```suggestion
    array, shape (2, 2)
```

ufmayer replied:
Done.

jnothman commented on sklearn/metrics/cluster/_supervised.py:
```diff
@@ -138,6 +146,206 @@ def contingency_matrix(labels_true, labels_pred, *, eps=None, sparse=False):
+
+    Examples
+    --------
+    Perfectly matching labelings have all non-zero entries on the
```
Please give just one or two examples here, not detailed instruction.

ufmayer replied:
Done.

jnothman commented on sklearn/metrics/cluster/_supervised.py:
```diff
@@ -138,6 +146,206 @@ def contingency_matrix(labels_true, labels_pred, *, eps=None, sparse=False):
+
+    Examples
+    --------
+    Perfectly matching labelings have a score of 1 even
```
Similarly, please reduce this

ufmayer replied:
Done.

jnothman commented on sklearn/metrics/cluster/_supervised.py:
```diff
@@ -138,6 +146,206 @@ def contingency_matrix(labels_true, labels_pred, *, eps=None, sparse=False):
+    assigned in the same or different clusters in the predicted and
+    true clusterings.
+
+    The raw RI score is:
```
Similarly remove this and reference the User Guide

ufmayer replied:
Done.

jnothman commented on sklearn/metrics/cluster/_supervised.py:
```diff
@@ -138,6 +146,206 @@ def contingency_matrix(labels_true, labels_pred, *, eps=None, sparse=False):
+    return m_diag / m_sum
+
+
 def adjusted_rand_score(labels_true, labels_pred):
```
Shouldn't we use `pair_confusion_matrix` in `adjusted_rand_score` too?

ufmayer replied:
I would leave it as is. The code currently reflects how the main wikipedia page on the rand index is written [https://en.wikipedia.org/wiki/Rand_index]. Both wikipedia and scikit have for the raw rand index the computation via the pair confusion matrix and for the adjusted rand score via the contingency table.

jnothman reviewed (commented):
Otherwise looking good, thanks :)

ufmayer committed:
Merge branch 'master' of https://github.com/scikit-learn/scikit-learn

ufmayer committed:
Reverting these changes, they were issues in master that have been patched there

ufmayer committed:
Simplified API docstrings

ufmayer commented on doc/modules/clustering.rst:
```diff
@@ -1043,6 +1043,117 @@ classes according to some similarity metric.
+  >>> metrics.rand_score(labels_true, labels_pred)
+  0.66...
+
+One can permute 0 and 1 in the predicted labels, rename 2 to 3, and get
```
Done.

ufmayer commented on doc/modules/clustering.rst:
```diff
@@ -1043,6 +1043,117 @@ classes according to some similarity metric.
+Advantages
+~~~~~~~~~~
+
+- **Interpretable**: Taking a clustering and all possible element
```
Done.

ufmayer commented:
@jnothman I incorporated your suggested changes. Also see the comment above on the computation of ARI.

jnothman commented:
Wikipedia is not a basis for standardisation.  A story: Once upon a time Wikipedia's article on Confusion Matrix matched our implementation in confusion_matrix with respect to which axis was which. Then someone changed Wikipedia and transposed its confusion matrices, and users started reporting that our implementation was incorrect or non-standard. Another story: assignments for a course I taught referred students to Wikipedia to help them implement Cohen's Kappa. Until one year I was lucky to look at Wikipedia on Cohen's Kappa before the students had had their assignments too long. Someone had edited it into an inconsistent and incomprehensible rubbish. I hurriedly patched it up.

Internal consistency and avoidance of redundant duplication is a much more valuable objective than matching Wikipedia.

ufmayer committed:
Fix typo in doc string of pair_confusion_matrix

ufmayer committed:
Force rand_score result to be float for python2

ufmayer commented:
I fully hear what you say, I edit Wikipedia at times as well. In this case I would say that Wikipedia reflects the usual approach to these two scores. Then again if you prefer I can certainly change the implementation of adjusted_rand_score to use pair_confusion_matrix. The code becomes neat and clean.

```
def adjusted_rand_score_pcm(labels_true, labels_pred):
    """Rand index adjusted for chance.
   ... documentation as is ...
    References
    ----------
    .. [Hubert1985] L. Hubert and P. Arabie, Comparing Partitions,
      Journal of Classification 1985
      https://link.springer.com/article/10.1007%2FBF01908075
      
    .. [Steinley2004] D. Steinley, Properties of the Hubert-Arabie
      adjusted Rand index, Psychological Methods 2004

    .. [wk] https://en.wikipedia.org/wiki/Rand_index#Adjusted_Rand_index

    See also
    --------
    adjusted_mutual_info_score: Adjusted Mutual Information

    """
    n = len(labels_true)
    m = pair_confusion_matrix(labels_true, labels_pred)
    a = m[1][1]
    b = m[1][0]
    c = m[0][1]
    d = m[0][0]

    # Special cases: empty data or full agreement
    if n == 0 or (b == 0 and c == 0):
        return 1.0

    return 1. - 2. * _comb2(n) * (b + c) / ((a + b) * (b + d) +
                                            (a + c) * (c + d))
```

I added the Steinley reference as the base of the formula, and left the doc string as it was otherwise. I am not sure which code will be easier to maintain because familiarity with the algorithms by the maintainers is part of that decision. You are in a much better place to judge that.

At this point I have not added this to the PR. Just let me know what you prefer and I'll add it if so desired.

jnothman commented:
I think that implementation of ARI makes the "comb" less mysterious.

jnothman commented on sklearn/metrics/cluster/_supervised.py:
```diff
@@ -138,6 +146,147 @@ def contingency_matrix(labels_true, labels_pred, *, eps=None, sparse=False):
+
+    Returns
+    -------
+    array, shape(2, 2)
```
we need to be clear on the semantics of each axis. They should correspond to confusion matrix: axis 0 should represent the true clustering, and axis 1 the predicted clustering

ufmayer replied:
- I used pair_confusion_matrix and your variables in adjusted_rand_index and made it part of the PR, and added that Steinley2004 reference both to the doc string and to the user guide.
- I updated the doc string and user guide for pair_confusion_matrix following the doc of confusion_matrix making the semantics clear.
- I went back to the original (and faster) implementation of pair_confusion_matrix without _comb2 as pair_confusion_matrix now was the only method in _supervised.py using _comb2, and I removed the comb / _comb2 methods altogether.

jnothman reviewed (commented):
I think this is a little clearer for ARI:
```py
    (tn, fp), (fn, tp) = pair_confusion_matrix(labels_true, labels_pred)

    # Special cases: empty data or full agreement
    if fn == 0 and fp == 0:
        return 1.0

    return 2. * (tp*tn - fn*fp) / ((tp + fn) * (fn + tn) + (tp + fp) * (fp + tn))
```

ufmayer committed:
Updated pair_confusion_matrix docs & var names

ufmayer committed:
Implement adjusted_rand_score using pair_confusion_matrix

ufmayer committed:
Implement pair_confusion_matrix without _comb2, remove _comb2 altogether

ufmayer committed:
Updated matrix math display in pair_confusion_matrix user guide

ufmayer committed:
Whitespace fix in user guide

ufmayer committed:
Minor wording changes in pair_confusion_matrix user guide

ufmayer committed:
Minor doc string changes

jnothman commented:
Thank you @ufmayer. Let's await another review.

ufmayer committed:
Merged in master from from the upstream repository

ufmayer commented:
@cmarmo Is this something you could review? @jnothman has approved the PR.

cmarmo commented:
@ogrisel or @jeremiedbb, perhaps you might want to have a look at this. Thanks.

cmarmo commented on doc/whats_new/v0.24.rst:
```diff
@@ -215,16 +215,23 @@ Changelog
-  :func:`metrics.classification_report` which was raising AttributeError
-  when called with `output_dict=True` for 0-length values.
-  :pr:`17777` by :user:`Shubhanshu Mishra <napsternxg>`
+- |Feature| :func:`metrics.plot_confusion_matrix` now supports making colorbar
+  optional in the matplotlib plot by setting colorbar=False. :pr:`17192` by
+  :user:`Avi Gupta <avigupta2612>`
```
I'm wondering if something went wrong during a previous synchronisation with upstream... This block was below...

ufmayer replied:
@cmarmo The `what's new` instructions say: Entries should be grouped by module (in alphabetic order) and prefixed with one of the labels: |MajorFeature|, |Feature|, |Efficiency|, |Enhancement|, |Fix| or |API| ... Entries should be ordered by those labels (e.g. |Fix| after |Efficiency|).

I remember resorting an entry to follow those instructions during a manual merge conflict resolve. Currently I see:
```
- |Feature| Add :func:`metrics.cluster.pair_confusion_matrix` ...
- |Feature| Added :func:`metrics.mean_absolute_percentage_error` ...
- |Feature| :func:`metrics.plot_confusion_matrix` ...
- |Feature| Add :func:`metrics.rand_score`  ...
- |Enhancement| ...
```

Seem to be in the desired order, assuming the "Feature" label of Avi Gupta's metrics.plot_confusion_matrix entry is correct. Then again to me the making a color bar optional might be more of an "Enhancement" vs a "Feature". In that case the label should be changed from "Feature" to "Enhancement" and that block should be moved back down. What do you suggest?

cmarmo replied:
You are right: the issue here is that those modifications are not related with your pull request. If you decide to fix them it could be suitable to do that in a separate PR.

ufmayer replied:
Ok, I'll move it back down and I'll do a separate PR after this one is merged in (to avoid a PR on top of a PR).

cmarmo commented on sklearn/metrics/cluster/_supervised.py:
```diff
@@ -113,14 +108,20 @@ def contingency_matrix(labels_true, labels_pred, *, eps=None, sparse=False):
 
         .. versionadded:: 0.18
 
+    dtype : numeric data type, optional.
+        Defaults to np.int, but see the notes on ``eps`` below.
```
Following the [guidelines for writing documentation](https://scikit-learn.org/stable/developers/contributing.html#guidelines-for-writing-documentation)
```suggestion
    dtype : numeric data type, default=np.int.
        See the notes on ``eps`` below.
```

cmarmo commented on sklearn/metrics/cluster/_supervised.py:
```diff
@@ -113,14 +108,20 @@ def contingency_matrix(labels_true, labels_pred, *, eps=None, sparse=False):
+    dtype : numeric data type, optional.
+        Defaults to np.int, but see the notes on ``eps`` below.
+
+        .. versionadded:: 0.22.2
```
```suggestion
        .. versionadded:: 0.24
```

cmarmo reviewed (commented):
@ufmayer thanks for your patience! Waiting for a core dev review, I've suggested some minor fixes. The ``what's new`` worth some attention during the sync with upstream.

ufmayer committed:
Correct versionadded

Co-authored-by: Chiara Marmo <cmarmo@users.noreply.github.com>

ufmayer committed:
Following the guidelines for writing documentation

Co-authored-by: Chiara Marmo <cmarmo@users.noreply.github.com>

ufmayer committed:
Merged in master from from the upstream repository

ufmayer committed:
Re-establish order of 0.24 changes in master

ufmayer committed:
Merged in master from from the upstream repository

ufmayer commented:
@jnothman @cmarmo It's been 3 months today since I filed this PR. Is there anything you could suggest to move this along?

jnothman commented:
I've tagged it for the current release milestone, and hopefully it will be picked up by another reviewer as we work towards release over the next 6 weeks.

Sorry for the delays, but there's a lot to review!

ufmayer commented:
@jnothman Thanks!

ufmayer committed:
Merged in master from from the upstream repository

ufmayer committed:
Merge branch 'master' of https://github.com/scikit-learn/scikit-learn into master

ufmayer committed:
Merged in master from from the upstream repository

glemaitre commented on sklearn/metrics/cluster/_supervised.py:
```diff
@@ -150,6 +151,153 @@ def contingency_matrix(labels_true, labels_pred, *, eps=None, sparse=False):
+
+    """
+    labels_true, labels_pred = check_clusterings(labels_true, labels_pred)
+    n_samples = np.int64(labels_true.shape[0])
```
```suggestion
    n_samples = labels_true.shape[0]
```

ufmayer replied:
I left this unchanged for now. The entire calculations below are using np.int64 and the return data type of the ndarray is np.int64, and this cast makes this explicitly clear. The cast will of course otherwise happen implicitly anyway but I think this makes it unambiguous. I am not opposed to the change, but if you don't feel strongly I would prefer to let it stand.

glemaitre commented on sklearn/metrics/cluster/_supervised.py:
```diff
@@ -150,6 +151,153 @@ def contingency_matrix(labels_true, labels_pred, *, eps=None, sparse=False):
+    n_samples = np.int64(labels_true.shape[0])
+
+    # Computation using the contingency data
+    contingency = contingency_matrix(labels_true, labels_pred, sparse=True,
+                                     dtype=np.int64)
```
```suggestion
    contingency = contingency_matrix(
        labels_true, labels_pred, sparse=True, dtype=np.int64
    )
```

glemaitre commented on sklearn/metrics/cluster/_supervised.py:
```diff
@@ -150,6 +151,153 @@ def contingency_matrix(labels_true, labels_pred, *, eps=None, sparse=False):
+                                     dtype=np.int64)
+    n_c = np.ravel(contingency.sum(axis=1))
+    n_k = np.ravel(contingency.sum(axis=0))
+    Sum_Squares = sum(n_ij*n_ij for n_ij in contingency.data)
```
```suggestion
    sum_squares = (contingency.data ** 2).sum()
```

glemaitre commented on sklearn/metrics/cluster/_supervised.py:
```diff
@@ -150,6 +151,153 @@ def contingency_matrix(labels_true, labels_pred, *, eps=None, sparse=False):
+    n_c = np.ravel(contingency.sum(axis=1))
+    n_k = np.ravel(contingency.sum(axis=0))
+    Sum_Squares = sum(n_ij*n_ij for n_ij in contingency.data)
+    C11 = Sum_Squares - n_samples
+    C01 = sum(contingency.dot(n_k)) - Sum_Squares
+    C10 = sum(contingency.transpose().dot(n_c)) - Sum_Squares
+    C00 = n_samples * n_samples - C01 - C10 - Sum_Squares
+    return np.array([[C00, C01], [C10, C11]])
```
```suggestion
    C = np.empty((2, 2), dtype=np.int64)
    C[1, 1] = sum_squares - n_samples
    C[0, 1] = contingency.dot(n_k).sum() - sum_squares
    C[1, 0] = contingency.transpose().dot(n_c).sum() - sum_squares
    C[0, 0] = n_samples **2  - C[0, 1] - C[1, 0] - sum_squares
    return C
```

glemaitre reviewed (commented):
First round of reviews.

glemaitre commented on sklearn/metrics/cluster/tests/test_supervised.py:
```diff
@@ -370,3 +379,70 @@ def test_check_clustering_error():
+
+    # regular case: different non-trivial clusterings
+    clustering1 = np.array([i+1 for i in range(n) for j in range(n)])
+    clustering2 = np.array([i+1 for i in range(n) for j in range(n+1)][:N])
```
```suggestion
    clustering2 = np.hstack([[i+1] * (n+1) for i in range(n)])[:N]
```

glemaitre commented on sklearn/metrics/cluster/tests/test_supervised.py:
```diff
@@ -370,3 +379,70 @@ def test_check_clustering_error():
+                                                       clustering2))
+
+    # regular case: different non-trivial clusterings
+    clustering1 = np.array([i+1 for i in range(n) for j in range(n)])
```
```suggestion
    clustering1 = np.hstack([[i+1] * n for i in range(n)])
```

glemaitre commented on sklearn/metrics/cluster/tests/test_supervised.py:
```diff
@@ -370,3 +379,70 @@ def test_check_clustering_error():
+                same_cluster_1 = int(clustering1[i] == clustering1[j])
+                same_cluster_2 = int(clustering2[i] == clustering2[j])
+                expected[same_cluster_1, same_cluster_2] += 1
+    assert_array_equal(expected, pair_confusion_matrix(clustering1,
+                                                       clustering2))
```
```suggestion
    assert_array_equal(
        expected, pair_confusion_matrix(clustering1, clustering2)
    )
```

ufmayer replied:
In general the numpy.testing API expects the arguments to be of the order (actual, desired, ...), so I changed this to:
```
    assert_array_equal(
        pair_confusion_matrix(clustering1, clustering2), expected)
    )
```
The same comment about the argument order applies to the other tests I implemented.

glemaitre replied:
Yes you are right. In testing the order is usually `assert computed == expected`.

glemaitre commented on sklearn/metrics/cluster/tests/test_supervised.py:
```diff
@@ -370,3 +379,70 @@ def test_check_clustering_error():
+    clustering1 = np.array([i+1 for i in range(n) for j in range(n)])
+    clustering2 = np.array([i+1 for i in range(n) for j in range(n+1)][:N])
+    # basic quadratic implementation
+    expected = np.full(shape=(2, 2), fill_value=0, dtype=np.int64)
```
```suggestion
    expected = np.zeros(shape=(2, 2), dtype=np.int64)
```

glemaitre commented on sklearn/metrics/cluster/tests/test_supervised.py:
```diff
@@ -370,3 +379,70 @@ def test_check_clustering_error():
+    n = 10
+    N = n * n
+
+    # edge case: every element is its own cluster
```
We could have separated tests for normal case and edge case. I think that we can parametrize the test:

```python
@pytest.mark.parametrize(
    "clustering1, clustering2",
    [(range(100), range(100)),
     (np.zeros((100,)), np.zeros((100,))]
)
def test_rand_score_edge_case(clustering1, clustering2):
    assert_array_almost_equal(
        1.0, rand_score(clustering1, clustering2)
    )
```

glemaitre replied:
we could have something similar for one the edges cases above

ufmayer replied:
Done.

glemaitre commented on sklearn/metrics/cluster/tests/test_supervised.py:
```diff
@@ -370,3 +379,70 @@ def test_check_clustering_error():
+    clustering1 = list(range(N))
+    clustering2 = clustering1
+    expected = np.array([1.])
+    assert_array_almost_equal(expected, rand_score(clustering1,
```
It would be best to use `assert_allclose` instead of `assert_array_almost_equal`

ufmayer replied:
Done.

glemaitre reviewed (commented):
Some remaining comments.
But in general it looks good.

ufmayer committed:
Implement changes suggested by reviewer

ufmayer committed:
Fix whitespace for linter

ufmayer committed:
Fix typo

ufmayer committed:
Fix doc formatting

ufmayer committed:
Remove tests replaced by parameterized implementation

ufmayer committed:
Combine unadjusted and adjusted Rand index doc

ufmayer commented on doc/modules/clustering.rst:
```diff
@@ -1043,6 +1043,115 @@ classes according to some similarity metric.
 
+.. _rand_score:
+
+Rand index
```
I merged the two sections. Please have a look.

ufmayer commented on sklearn/metrics/cluster/_supervised.py:
```diff
@@ -113,14 +108,20 @@ def contingency_matrix(labels_true, labels_pred, *, eps=None, sparse=False):
 
         .. versionadded:: 0.18
 
+    dtype : numeric data type, default=np.int.
```
I changed it to `dtype : numeric type, default=np.int` as it gives additional information that the dtype cannot be arbitrary.

ufmayer commented on sklearn/metrics/cluster/_supervised.py:
```diff
@@ -113,14 +108,20 @@ def contingency_matrix(labels_true, labels_pred, *, eps=None, sparse=False):
         .. versionadded:: 0.18
 
+    dtype : numeric data type, default=np.int.
+        See the notes on ``eps`` below.
```
Makes it a bit long, and duplicates the text, meaning if there is a change in the future on the `eps` explanation it would have to be made twice. I would prefer to leave it as is.

glemaitre replied:
Can we specify the following instead (if I don't make a mistake interpreting the note below).
```
Output dtype. Ignore if `eps` is not `None`.
```

ufmayer replied:
I added (note the extra "d"):
```
Output dtype. Ignored if `eps` is not `None`.
```

ufmayer commented on sklearn/metrics/cluster/_supervised.py:
```diff
@@ -150,6 +151,153 @@ def contingency_matrix(labels_true, labels_pred, *, eps=None, sparse=False):
+    -------
+    C : array, shape(2, 2)
+
+    Examples
```
I re-arranged to the following order which seems to be common in the overall codebase (which is inconsistent overall):
See Also
Examples
References

You may want to do this to see the inconsistencies in the codebase:
```
cd scikit-learn/sklearn/metrics/cluster
grep -P '(^def )|^\s+(Parameter|Return|See Also|Example|Reference)s?\s*$' *.py
```

ufmayer commented on sklearn/metrics/cluster/_supervised.py:
```diff
@@ -150,6 +151,153 @@ def contingency_matrix(labels_true, labels_pred, *, eps=None, sparse=False):
+    if m_diag == m_sum or m_sum == 0:
+        return 1.0
+
+    return float(m_diag) / m_sum
```
I left the cast to float for backward compatibility with python2. If that's not needed the cast to float can be removed. @glemaitre Please let me know.

glemaitre replied:
We don't support Python 2 that's actually the reason for my proposal :)

ufmayer replied:
OK.

ufmayer commented on sklearn/metrics/cluster/tests/test_supervised.py:
```diff
@@ -370,3 +379,70 @@ def test_check_clustering_error():
+    N = n * n
+
+    # edge case: every element is its own cluster
+    clustering1 = list(range(N))
```
I left the list() call because the API of pair_confusion_matrix expects an array-like parameter while range(N) is an iterable.

ufmayer commented on sklearn/metrics/cluster/tests/test_supervised.py:
```diff
@@ -370,3 +379,70 @@ def test_check_clustering_error():
+
+def test_pair_confusion_matrix():
+    n = 10
+    N = n * n
```
I made this
```
n = 10
N = n ** 2
```
because this makes the dependency of n and N clear. I prefer this to n,N = 10,100 because that might leave the impression that n and N can be set independently.

ufmayer commented:
@glemaitre Thank you for the thorough review. I implemented the majority of your suggestions and resolved the corresponding auto-generated conversations in this PR. Please have a look at the remaining ones, I left brief comments on each that I didn't resolve.

glemaitre commented on doc/modules/clustering.rst:
```diff
@@ -1043,86 +1043,117 @@ classes according to some similarity metric.
+``labels_true`` and our clustering algorithm assignments of the same
+samples ``labels_pred``, the **(adjusted or unadjusted) Rand index**
+is a function that measures the **similarity** of the two assignments,
+ignoring permutations (and for the adjusted Rand index **with chance
+normalization**)::
```
```suggestion
ignoring permutations.::
```

glemaitre commented on doc/modules/clustering.rst:
```diff
@@ -1043,86 +1043,117 @@ classes according to some similarity metric.
   >>> labels_pred = [0, 0, 1, 1, 2, 2]
-
+  >>> metrics.rand_score(labels_true, labels_pred)
+  0.66...
```
```suggestion
  0.66...

The Rand index does not ensure to obtain a value close to 0.0 for a
random labelling. The adjusted Rand index **correct for chance** and
will give such a baseline.
```

glemaitre commented on sklearn/metrics/cluster/_supervised.py:
```diff
@@ -113,14 +108,20 @@ def contingency_matrix(labels_true, labels_pred, *, eps=None, sparse=False):
 
         .. versionadded:: 0.18
 
+    dtype : numeric type, default=np.int.
```
```suggestion
    dtype : numeric type, default=np.int
```

glemaitre commented on sklearn/metrics/cluster/_supervised.py:
```diff
@@ -150,6 +151,156 @@ def contingency_matrix(labels_true, labels_pred, *, eps=None, sparse=False):
+    .. L. Hubert and P. Arabie, Comparing Partitions, Journal of
+      Classification 1985
+      https://link.springer.com/article/10.1007%2FBF01908075
+
```
Numpydoc specify to not have a blank line at the end of the block

```suggestion
```

glemaitre commented on sklearn/metrics/cluster/_supervised.py:
```diff
@@ -150,6 +151,156 @@ def contingency_matrix(labels_true, labels_pred, *, eps=None, sparse=False):
+    .. https://en.wikipedia.org/wiki/Simple_matching_coefficient
+
+    .. https://en.wikipedia.org/wiki/Rand_index
+
```
```suggestion
```

glemaitre reviewed (commented):
There is the remaining comments. Just some nitpicks. Once done this is mergeable :)

ufmayer committed:
Merging in latest changes requested by reviewers

ufmayer commented:
@glemaitre Done.

cmarmo commented:
@glemaitre, ready for a second approval? Thanks!

glemaitre commented:
Thanks @ufmayer LGTM merging.
